Source/WebCore/page/Navigation.cpp

```diff
@@ -161,23 +193,26 @@ Navigation::Result Navigation::navigate(ScriptExecutionContext& scriptExecutionC
     if (options.history == HistoryBehavior::Push && newURL.protocolIsJavaScript())
         return createErrorResult(committed, finished, ExceptionCode::NotSupportedError, "A \"push\" navigation was explicitly requested, but only a \"replace\" navigation is possible when navigating to a javascript: URL."_s);
 
-    if (options.state) {
-        Vector<RefPtr<MessagePort>> dummyPorts;
-        auto serializeResult = SerializedScriptValue::create(*scriptExecutionContext.globalObject(), options.state, { }, dummyPorts, SerializationForStorage::Yes);
-        if (serializeResult.hasException())
-            return createErrorResult(committed, finished, serializeResult.releaseException());
-    }
+    auto serializedState = serializeState(options.state);
+    if (serializedState.hasException())
+        return createErrorResult(committed, finished, serializedState.releaseException());
 
     if (!window()->frame() || !window()->frame()->document())
         return createErrorResult(committed, finished, ExceptionCode::InvalidStateError, "Invalid state"_s);
 
+    Ref apiMethodTracker = maybeSetUpcomingNonTraversalTracker(committed, finished, WTFMove(options.info), serializedState.releaseReturnValue());
+
     // FIXME: This is not a proper Navigation API initiated traversal, just a simple load for now.
     window()->frame()->loader().load(FrameLoadRequest(*window()->frame(), newURL));
 
+    if (apiMethodTracker == m_upcomingNonTraverseMethodTracker) {
+        // TODO: Once the frameloader properly promotes the upcoming tracker with the navigate event `m_upcomingNonTraverseMethodTracker` should be unset or this will throw.
+        m_upcomingNonTraverseMethodTracker = std::nullopt;
+    }
+
     // FIXME: keep track of promises to resolve later.
```

Comment:
Is this the job of NavigationAPIMethodTracker? If so I think the FIXME can be removed.

Reply:
Yes but not everything is done to resolve the promises at the right moment yet.